include/material_models/PseudoPlastic.h

```diff
@@ -0,0 +1,185 @@
+
+    void postprocess(Solver<3> &solver, Reader &reader, const char *resultsFileName, int load_idx, int time_idx) override
+    {
+        VectorXf element_plastic_flag = VectorXf::Zero(solver.local_n0 * solver.n_y * solver.n_z);
+        for (ptrdiff_t elem_idx = 0; elem_idx < solver.local_n0 * solver.n_y * solver.n_z; ++elem_idx) {
+            element_plastic_flag(elem_idx) = plastic_flag[elem_idx].cast<float>().mean();
+        }
+
+        if (find(reader.resultsToWrite.begin(), reader.resultsToWrite.end(), "plastic_flag") != reader.resultsToWrite.end()) {
+            for (int i = 0; i < solver.world_size; ++i) {
+                if (i == solver.world_rank) {
+                    char name[5096];
+                    sprintf(name, "%s/load%i/time_step%i/plastic_flag", reader.ms_datasetname, load_idx, time_idx);
+                    reader.WriteSlab<float>(element_plastic_flag.data(), 1, resultsFileName, name);
+                }
+                MPI_Barrier(MPI_COMM_WORLD);
+            }
+        }
+    }
+
+  protected:
+    vector<double>       bulk_modulus;
+    vector<double>       shear_modulus;
+    vector<double>       yield_stress;
+    vector<double>       eps_crit;
+    vector<VectorXi>     plastic_flag;
+    Matrix<double, 6, 1> dev_eps;
+    double               treps, norm_dev_eps, buf1, buf2;
+};
+
+class PseudoPlasticLinearHardening : public PseudoPlastic {
+  public:
+    PseudoPlasticLinearHardening(vector<double> l_e, map<string, vector<double>> materialProperties)
+        : PseudoPlastic(l_e, materialProperties)
+    {
+        try {
+            hardening_parameter = materialProperties["hardening_parameter"];
+        } catch (const std::exception &e) {
+            throw std::runtime_error("Missing material properties for the requested material model.");
+        }
+
+        E_s.resize(n_mat);
+        eps_crit.resize(n_mat);
+
+        for (int i = 0; i < n_mat; ++i) {
+            eps_crit[i] = sqrt(2. / 3.) * yield_stress[i] / (2. * shear_modulus[i]);
+            E_s[i]      = (3. * shear_modulus[i]) / (3. * shear_modulus[i] + hardening_parameter[i]);
+        }
+    }
+
+    void get_sigma(int i, int mat_index, ptrdiff_t element_idx) override
+    {
+        treps             = eps.block<3, 1>(i, 0).sum();
+        dev_eps.head<3>() = eps.block<3, 1>(i, 0) - (1.0 / 3.0) * treps * Matrix<double, 3, 1>::Ones();
+        dev_eps.tail<3>() = eps.block<3, 1>(i + 3, 0);
+
+        norm_dev_eps = dev_eps.norm();
+        buf1         = bulk_modulus[mat_index] * treps;
+
+        if (norm_dev_eps <= eps_crit[mat_index]) {
+            buf2                                 = 2.0 * shear_modulus[mat_index];
+            plastic_flag[element_idx](i / n_str) = mat_index;
+        } else {
+            buf2 = (b * yield_stress[mat_index] + a * E_s[mat_index] * hardening_parameter[mat_index] *
+                                                      (norm_dev_eps - eps_crit[mat_index])) /
+                   norm_dev_eps;
```

Comment:
Do we need a fallback for `norm_dev_eps` close to zero?

Reply:
if `norm_dev_eps` is close to zero (line 114: `if (norm_dev_eps <= eps_crit[mat_index])`), it will be in the linear elastic regime.